File: epos_restaurant_2023/gym/report/expired_member_report/expired_member_report.py
# Copyright (c) 2024, Tes Pheakdey and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import datetime, timedelta,date

def execute(filters=None):
	return get_columns(filters), get_data(filters)

def get_data(filters):
	start_date, end_date = get_date_range(filters)
	filter  = "a.docstatus = 1"
	filter += " AND a.end_date between '{0}' and '{1}'".format(start_date, end_date)
	sql = """
			select 
				a.customer,
				a.member_name,
				c.customer_group,
				a.gender,
				a.start_date,
				a.end_date,
				a.membership
			from `tabMembership` a
			inner join `tabCustomer` c on a.customer = c.name
			WHERE {0}
			""".format(filter)
	data = frappe.db.sql(sql,as_dict=1)
	return data

def get_date_range(filters):
	today = datetime.now()
	start_date = datetime.now().strftime("%Y-%m-%d")
	end_date = datetime.now().strftime("%Y-%m-%d")
	if filters.filter_based_on == "Today":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = datetime.now().strftime("%Y-%m-%d")
	elif filters.filter_based_on == "7 Days":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=7)).strftime("%Y-%m-%d")
	elif filters.filter_based_on == "15 Days":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=15)).strftime("%Y-%m-%d")
	elif filters.filter_based_on == "30 Days":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=30)).strftime("%Y-%m-%d")
	elif filters.filter_based_on == "3 Months":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=90)).strftime("%Y-%m-%d")
	elif filters.filter_based_on == "6 Months":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=180)).strftime("%Y-%m-%d")
	elif filters.filter_based_on == "9 Months":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=270)).strftime("%Y-%m-%d")
	elif filters.filter_based_on == "12 Months":
		start_date = datetime.now().strftime("%Y-%m-%d")
		end_date = (date.today() + timedelta(days=365)).strftime("%Y-%m-%d")
	else:
		start_date = filters.start_date
		end_date = filters.end_date
	return start_date, end_date

def get_columns(filters):
	columns = [
		{
			"label": _("ID"),
			"fieldname": "customer",
			"fieldtype": "Link",
			"options":"Customer",
			"width": 100,
			"align": "center"
		},
		{
			"label": _("Member Name"),
			"fieldname": "member_name",
			"fieldtype": "Data",
			"width": 200,
			"align": "left"
		},
		{
			"label": _("Member Type"),
			"fieldname": "customer_group",
			"fieldtype": "Data",
			"width": 200,
			"align": "left"
		},
		{
			"label": _("Gender"),
			"fieldname": "gender",
			"fieldtype": "Data",
			"width": 100,
			"align": "center"
		},
		{
			"label": _("Membership Option"),
			"fieldname": "membership",
			"fieldtype": "Data",
			"width": 250,
			"align": "center"
		},
		{
			"label": _("Start Date"),
			"fieldname": "start_date",
			"fieldtype": "Data",
			"width": 120,
			"align": "center"
		},
		{
			"label": "Expiry Date",
			"fieldname": "end_date",
			"fieldtype": "Data",
			"width": 120,
			"align": "center"
		},
	]
	
	return columns

def get_list(filters,name):
	data = ','.join("'{0}'".format(x.replace("'", "''")) for x in filters.get(name))
	return data

File: epos_restaurant_2023/gym/report/expired_member_report/expired_member_report.js
// Copyright (c) 2026, Tes Pheakdey and contributors
// For license information, please see license.txt

frappe.query_reports["Expired Member Report"] = {
	onload: function (report) {
		report.page.add_inner_button("Preview Report", function () {
			frappe.query_report.refresh();
		});
		if(frappe.query_report.get_filter_value('filter_based_on') != "Custom"){
			frappe.query_report.toggle_filter_display('start_date', true  );
			frappe.query_report.toggle_filter_display('end_date', true );
		}
		const custom_css = `
            <style>
                .datatable .dt-row .dt-cell:first-child .dt-cell__content{
					min-width: 100px !important;
					width: 100px !important;
					align-content: center !important;
				}
            </style>`;
        $(custom_css).appendTo("head");
	},
	"filters": [
		{
			"fieldname":"filter_based_on",
			"label": __("Based On"),
			"fieldtype": "Select",
			"reqd": 1,
			"options": ["Today", "7 Days", "15 Days","30 Days","3 Months","6 Months","9 Months","12 Months","Custom"],
			"default": "Today",
			"on_change": function (query_report) {
				let filter_based_on = frappe.query_report.get_filter_value('filter_based_on');
				if(filter_based_on === "Custom"){
					frappe.query_report.toggle_filter_display('start_date', false  );
					frappe.query_report.toggle_filter_display('end_date', false );
				}
				else{
					frappe.query_report.toggle_filter_display('start_date', true  );
					frappe.query_report.toggle_filter_display('end_date', true );
				}
			},
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			default:frappe.datetime.get_today(),
			"reqd": 1,
			"on_change": function (query_report) {},
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			default:frappe.datetime.get_today(),
			"reqd": 1,
			"on_change": function (query_report) {},
		},
	]
};
